include/wil/safearrays.h

```diff
@@ -0,0 +1,1105 @@
+                    }
+                }
+                *result = static_cast<ULONG>(elements);
+            }
+            else
+            {
+                // If it's invalid, it doesn't contain any elements
+                *result = 0;
+            }
+            return S_OK;
+        }
+
+        typedef resource_policy<SAFEARRAY*, decltype(&details::SafeArrayDestroy), details::SafeArrayDestroy, details::pointer_access_all> safearray_resource_policy;
+    }
+    /// @endcond
+
+    typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnlock), details::SafeArrayUnlock, details::pointer_access_noaddress> safearray_unlock_scope_exit;
+    typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnaccessData), details::SafeArrayUnaccessData, details::pointer_access_noaddress> safearray_unaccess_scope_exit;
+
+    //! Guarantees a SafeArrayUnlock call on the given object when the returned object goes out of scope
+    //! Note: call SafeArrayUnlock early with the reset() method on the returned object or abort the call with the release() method
+    WI_NODISCARD inline safearray_unlock_scope_exit SafeArrayUnlock_scope_exit(SAFEARRAY* psa) WI_NOEXCEPT
+    {
+        details::SafeArrayLock(psa);
+        return safearray_unlock_scope_exit(psa);
+    }
+
+    //! Guarantees a SafeArrayUnaccessData call on the given object when the returned object goes out of scope
+    //! Note: call SafeArrayUnaccessData early with the reset() method on the returned object or abort the call with the release() method
+    template<typename T>
+    WI_NODISCARD inline safearray_unaccess_scope_exit SafeArrayAccessData_scope_exit(SAFEARRAY* psa, T*& p) WI_NOEXCEPT
+    {
+        details::SafeArrayAccessData(psa, p);
+        return safearray_unaccess_scope_exit(psa);
+    }
+
+    //! Class that facilitates the direct access to the contents of a SAFEARRAY and "unaccessing" it when done
+    //! It enables treating the safearray like an regular array and/or performing a ranged-for on the contents.
+    //! Accessing a safearray increases its lock count, so attempts to destroy the safearray will fail until it is unaccessed.
+    //! This class works even if the SAFEARRAY is multi-dimensional by treating it as a large single-dimension array,
+    //! but doesn't support access via a multi-dimensional index.
+    //! NOTE: This class does not manage the lifetime of the SAFEARRAY itself.  See @ref safearray_t.
+    //! ~~~~
+    //! HRESULT copy_to_bstr_vector(SAFEARRAY* psa, std::vector<wil::unique_bstr>& result)
+    //! {
+    //!     auto data = wil::safearraydata_nothrow<BSTR>{};
+    //!     RETURN_IF_FAILED(data.access(psa));
+    //!     result.reserve(result.size() + data.size());
+    //!     for (BSTR& bstr : data)
+    //!     {
+    //!         result.emplace_back(wil::make_bstr_nothrow(bstr));
+    //!         RETURN_HR_IF_NULL(E_OUTOFMEMORY, result.rbegin()->get());
+    //!     }
+    //!     return S_OK;
+    //! }
+    //! ~~~~

```

Comment:
Does Doxygen handle this correctly? I'd think you'd need another set of `~~~` indicating the start of a new example